MSC2946: Spaces Summary (#2946)
matrix-org/matrix-doc: proposals/2946-spaces-summary.md

### Spaces Summary API

*This MSC depends on [MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772).*

Spaces are rooms with `m.space` as the [room type](https://github.com/matrix-org/matrix-doc/pull/1840).
Spaces can include state events to specify parent/child relationships.
These relationships point to other rooms, which may themselves be spaces.
This means spaces can have subspaces and rooms. This creates a graph: a space directory.

> This MSC defines a new endpoint which can be used to reveal information about the space directory.

Reviewer: ```suggestion
This MSC defines a new endpoint which can be used to reveal information about the space hierarchy.
```

imo "directory" implies that we're talking about /publicRooms or something, which we're not

Reviewer 2: Hierarchy is a misnomer too though as it implies there's a root, which there really isn't.

Reviewer 3: But doesn't the retrieved room tree have the queried room as its root? From my understanding, while there might not be a single universal root, the query response is indeed a hierarchy with the queried room as its root. In that case, why not say it as `This MSC defines a new endpoint which can be used to reveal information about the space hierarchy under a given space`?

Reviewer 2: > But doesn't the retrieved room tree have the queried room as its root?

Not really. This API just exposes nodes/edges, so a graph like:
```
  A
  |
  B
  |
  C   D
  |___|
    |
    E
```
with root room = D will return all the way up to A. The root room ID is just where to start exploring the graph.

     * `room_type` which is the room type, which is `m.space` for subspaces. It can be omitted if there is no room type
       in which case it should be interpreted as a normal room.
 - `events`: These are the edges of the graph. The objects in the array are complete (or stripped?) `m.room.parent`
   or `m.space.child` events.
 - `next_batch`: Its presence indicates that there are more results to return.

Server behaviour:
 - Extract the room ID from the request. Sanity check request data. Begin walking the graph
   starting with the room ID in the request in a queue of unvisited rooms according to the
   following rules:
    * If this room has already been processed, skip. NB: do not remember this between calls,
      as servers will need to visit the same room more than once to return additional events.
    * Is the caller currently joined to the room or is the room `world_readable`?
      If no, skip this room. If yes, continue.
    * If this room has not ever been in `rooms` (across multiple requests), extract the
      `PublicRoomsChunk` for this room.
    * Get all `m.space.child` and `m.room.parent` state events for the room. *In addition*, get

>       all `m.space.child` and `m.room.parent` state events which *point to* (via `state_key` or `content.room_id`)

Reviewer 4: This should probably factor in https://github.com/matrix-org/matrix-doc/pull/1772/files#diff-cdaa2b5bfe73f40db1903d17b1d242d8caa54d4af81e78e0396a369bcd9a1b90R140-R143 as it has the potential to cut trees shorter and further alleviate work off clients.

Reviewer 2: The API needs to support it so we can track secret rooms. It's not that onerous for clients to check there is both a parent->child and child->parent relationship if they only want to render bi-directional links. History visibility checks are done so it'll only actually return these non-bi-directional links if you are joined to a room which specifies it already, so I don't think this is a spam vector.

Reviewer 4: I don't think that quite covers it, its to prevent Joe Bloggs from adding a "Secret Room" to "Official Community" when he has absolute no permissions in "Official Community" - he shouldn't be able to make it show up in their hierarchy

Reviewer 2: It wouldn't unless you were already a part of "Official Community".

### Spaces Summary API

*This MSC depends on [MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772).*

> Spaces are rooms with `m.space` as the [room type](https://github.com/matrix-org/matrix-doc/pull/1840).

Reviewer 5: Is this an intentional departure from using a field in the `m.room.create` [we landed on in MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772/files/6cc3995ce825e5afbf29272967aef8adde899597#r526747733)? Just asking because this is not what the room type MSC is currently proposing.

Reviewer 2: No, the intention is to follow what MSC1772 says.

Reviewer 6: please can it be made to follow what MSC1772 says then?

```
POST /_matrix/client/r0/rooms/{roomID}/spaces
{
    "max_rooms_per_space": 5,      // The maximum number of rooms/subspaces to return for a given space, if negative unbounded. default: -1.
    "limit": 100,                  // The maximum number of rooms/subspaces to return, server can override this, default: 100.
    "batch": "opaque_string"       // A token to use if this is a subsequent HTTP hit, default: "".
}
```

which returns:

```
{
    "next_batch": "opaque string",
    "rooms": [
        {
            "aliases": [
                "#murrays:cheese.bar"
            ],
            "avatar_url": "mxc://bleeker.street/CHEDDARandBRIE",
            "guest_can_join": false,
            "name": "CHEESE",
            "num_joined_members": 37,
            "room_id": "!ol19s:bleecker.street",
            "topic": "Tasty tasty cheese",
            "world_readable": true,

            "num_refs": 42,
            "room_type": "m.space"
        },
        { ... }
    ],
    "events": [
        {
            "type": "m.space.child",
            "state_key": "!efgh:example.com",
            "content": {
                "via": ["example.com"],
                "present": true,

>                 "order": "abcd",

Reviewer 4: Will the server be respecting this order too?

Reviewer 2: I've purposefully left `events` as unsorted for now, and it's undefined which edges get returned first. We can use the `order` key but that alone is insufficient to get a stable sort (nothing stops the same order being present in different rooms, so which comes first?). We need to figure out a tie break mechanism.

> * Letting users classify their rooms: for example, separating "work" from "personal" rooms.
>
> We refer to such collections of rooms as "spaces".

This MSC attempts to solve how a member of a space discovers rooms in that space. This
is useful for quickly exposing a user to many aspects of an entire community, using the
examples above, joining the "official matrix.org rooms" space might suggest joining a few
rooms:

* A room to discuss development of the Matrix Spec.
* An announcements room for news related to matrix.org.
* An off-topic room for members of the space.

## Proposal

A new client-server API (and corresponding server-server API) is added which allows
for querying for the rooms and spaces contained within a space. This allows a client
to efficiently display a hierarchy of rooms to a user (i.e. without having
to walk the full state of the space).

### Client-server API

An endpoint is provided to walk the space tree, starting at the provided room ID
("the root room"), and visiting other rooms/spaces found via `m.space.child`
events. It recurses into the children and into their children, etc.

Example request:

```jsonc
POST /_matrix/client/r0/rooms/{roomID}/spaces

> 

Reviewer 7: If there's no `via` here, is it required that the user has already joined the root room?

Reviewer 8: The user must have joined the root room or it must be world readable.

Reviewer 8: I've clarified this in the errors section I believe.

* **`max_rooms_per_space`**: Optional: a client-defined limit to the maximum
  number of children to return per space. Doesn't apply to the root space (ie,
  the `room_id` in the request).

  Server implementations may also have an internal limit (recommended as 50)
  (which *does* apply to the root room); attempts to exceed this  limit are
  ignored. Must be a non-negative integer.

Response fields:

* **`rooms`**: for each room/space, starting with the root room, a
  summary of that room. The fields are the same as those returned by
  `/publicRooms` (see
  [spec](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-client-r0-publicrooms)),
  with the addition of:
  * **`room_type`**: the value of the `m.type` field from the
    room's `m.room.create` event, if any.
  * **`creation_ts`**: the value of the `origin_server_ts` field from the
    room's `m.room.create` event. This is required for sorting of rooms as specified
    in [MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772).
* **`events`**: `m.space.child` events of the returned rooms. For each event, only the
  following fields are returned: `type`, `state_key`, `content`, `room_id`,
  `sender`.

>   

Reviewer 8: There's currently no requirement that the room is a space, it just looks at whether there's any `m.space.child` events in it.

We should document what happens if the room doesn't exist / if you don't have permission to view it though! 👍

>   

Reviewer 8: I've clarified this by moving around some comments.

#### Algorithm

1. Start at the "root" room (the provided room ID).
2. Generate a summary and add it to `rooms`.
3. Add any `m.space.child` events in the room to `events`.
4. Recurse into the targets of the `m.space.child` events, generate a summary for

>    each room and add it to `rooms`, also and any `m.space.child` events of the room

Reviewer 7: ```suggestion
   each room and add it to `rooms`, also add any `m.space.child` events of the room
```

?

   to `events`.
5. Recurse into grandchildren, etc. until either all discovered rooms have been

>    inspected, or the  server-side limit on the number of rooms is reached.

Reviewer 7: You've mentioned below that clients must be aware of cycles but if you're specifying an exact algorithm here then I think, ironically, you're missing a cycle-breaking algorithm on the server side here 😭 Also, from reading I'm not 100% sure if this is breadth-first or depth-first. Having this algorithm here really helps with understanding though, so I think it would be fine to say this is a rough guide and the server impl can decide on the specifics (or is it important that all HSes generate exactly the same order?)

Reviewer 8: It isn't meant to be an exact algorithm. Synapse currently will not revisit rooms, but has no specific algorithm for cycle-breaking. (Note that the `rooms` and `events` returned are unordered.)

We could re-word this into a paragraph, which is how @richvdh originally had it, but I found it easier as a list of steps. Adding some words around it saying that this is the vague pseudo-code might be good.

I don't think the ordering matters since the returned data is unordered. It is also potentially problematic that the cut-off for what is returned could be in different places (if one implementation uses depth-first and one uses breadth-first they could cut-off at different locations if the limit is reached).

Reviewer 8: Oh, hmm...there's a note below that says:

> The current implementation doesn't honour `order` fields in child events, as
> suggested in [MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772).

I guess we want this to be stable so we should probably implement that.

Reviewer 8: I made the cycle breaking a bit clearer, but maybe we should specify breadth vs depth first search?

  that room does not exist), a 403 error is returned with `M_FORBIDDEN`. Any
  inaccessible children are simply omitted from the result (though the `m.space.child`
  events that point to them are still returned).
* There could be loops in the returned child events - clients should handle this
  gracefully.
* Similarly, note that a child room might appear multiple times (e.g. also be a
  grandchild).
* `suggested_only` applies transitively.

  For example, if a space A has child space B which is *not* suggested, and space
  B has suggested child room C, and the client makes a summary request for A with
  `suggested_only=true`,  neither B **nor** C will be returned.

  Similarly, if a space A has child space B which is suggested, and space B has
  suggested child room C which is suggested, and the client makes a summary request
  for A with `suggested_only=true`, both B and C will be returned.
* The current implementation doesn't honour `order` fields in child events, as
  suggested in [MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772).
* `m.space.child` with an invalid `via` (invalid is defined as missing, not an
  array or an empty array) are ignored.

> ### Server-server API

Reviewer 7: I think I'm missing how the server-server API fits into this: I'd assumed the HS would go & query the via servers for any rooms in finds that it doesn't already know about, but the presence of a server-server API suggests not. Is the idea to instead use this API on one of the via servers and splice the results into its own results?

Reviewer 8: Your assumption is pretty much correct, but the server must know about the root room, if it doesn't than an error is returned. This is not clear (and maybe it should take the `via` from the client to request info on it if it doesn't know it).

While visiting nodes in the tree, if the server does not know about a room it will reach out over federation (since each node includes `via` servers) and then include those results in the response.

Reviewer 8: I added some more information about this, hopefully it makes sense now!

  from the response.

This is largely the same as the Client-Server API, but differences are:

* The calling server can specify a list of spaces/rooms to omit from the
  response (via `exclude_rooms`).
* `max_rooms_per_space` applies to the root room as well as any returned
  children.
* If the target server is not a member of any discovered children (so
  would have to send another request over federation to inspect them), no
  attempt is made to recurse into them - they are simply omitted from the
  response.
  * If the target server is not a member of the root room, an empty
    response is returned.
* Currently, no consideration is given to room membership: the spaces/rooms
  must be world-readable (ie, peekable) for them to appear in the results.

## Potential issues

To reduce complexity, only a limited number of rooms are returned for a room,
no effort is made to paginate the results. Proper pagination is left to a future
MSC.

## Alternatives

An initial version of this followed both `m.space.child` and `m.space.parent` events,
but this is unnecessary to provide the expected user experience.

## Security considerations

> None.

Reviewer 7: Does it make it any easier to cause an HS to make a lot of outbound queries if I create a space with a lot of sub-spaces with the same / different HSes in the `via` param maybe?

Reviewer 8: Yeah this would be a DoS vector, good call! Especially since there's no guarantee the room IDs are actually valid.

Reviewer 8: I added some brief notes. I think this is mitigated a bit by limiting the total number of rooms visited. Maybe we should rate limit the endpoint though...

## MSC2946: Spaces Summary

This MSC depends on [MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772), which
describes why a Space is useful:

> Collecting rooms together into groups is useful for a number of purposes. Examples include:
>
> * Allowing users to discover different rooms related to a particular topic: for example "official matrix.org rooms".
> * Allowing administrators to manage permissions across a number of rooms: for example "a new employee has joined my company and needs access to all of our rooms".
> * Letting users classify their rooms: for example, separating "work" from "personal" rooms.
>
> We refer to such collections of rooms as "spaces".

> MSC2946 attempts to solve how a member of a space discovers rooms in that space. This

Reviewer 9: nitpick: when referencing the current MSC, don't do so by number.

```suggestion
This MSC attempts to solve how a member of a space discovers rooms in that space. This
```

            },
            "room_id": "!ol19s:bleecker.street",
            "sender": "@alice:bleecker.street"
        },
        { ... }
    ]
}
```

Request params:

* **`suggested_only`**: Optional. If `true`, return only child events and rooms where the
  `m.space.child` event has `suggested: true`. Defaults to `false`.

  For the POST request, must be a boolean. For GET, must be either `true` or `false`,
  case sensitive.
* **`max_rooms_per_space`**: Optional: a client-defined limit to the maximum
  number of children to return per space. Doesn't apply to the root space (ie,
  the `room_id` in the request).

  Server implementations may also have an internal limit (recommended to be 50)
  (which *does* apply to the root room); attempts to exceed this  limit are
  ignored. Must be a non-negative integer.

Response fields:

* **`rooms`**: for each room/space, starting with the root room, a
  summary of that room. The fields are the same as those returned by
  `/publicRooms` (see
  [spec](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-client-r0-publicrooms)),
  with the addition of:
  * **`room_type`**: the value of the `m.type` field from the

>     room's `m.room.create` event, if any.

Reviewer 4: Sorting children within a space requires the `origin_server_ts` of the child's `m.room.create` event, given we don't get the `m.room.create` as the room type is aggregated into the `rooms` array the ts of that event should be too

https://github.com/matrix-org/matrix-doc/pull/1772/files#diff-cdaa2b5bfe73f40db1903d17b1d242d8caa54d4af81e78e0396a369bcd9a1b90R122

Reviewer 8: This got removed from MSC1772, so not sure there's anything to include here.

Reviewer 4: It wasn't removed? https://github.com/matrix-org/matrix-doc/pull/1772/files#diff-cdaa2b5bfe73f40db1903d17b1d242d8caa54d4af81e78e0396a369bcd9a1b90R118-R127

Reviewer 8: Oh, my mistake. I thought the ordering was needed to break cycles, but it is just used generally. It should be easy enough to include, yes.

Reviewer 8: I've pushed f145fa3e8ac63edb5d7e9777cb55ecefea20beb8 with this change (and will work on getting it into Synapse).

Reviewer 8: See matrix-org/synapse#9928.

examples above, joining the "official matrix.org rooms" space might suggest joining a few
rooms:

* A room to discuss development of the Matrix Spec.
* An announcements room for news related to matrix.org.
* An off-topic room for members of the space.

## Proposal

A new client-server API (and corresponding server-server API) is added which allows
for querying for the rooms and spaces contained within a space. This allows a client
to efficiently display a hierarchy of rooms to a user (i.e. without having
to walk the full state of the space).

### Client-server API

An endpoint is provided to walk the space tree, starting at the provided room ID
("the root room"), and visiting other rooms/spaces found via `m.space.child`
events. It recurses into the children and into their children, etc.

Note that there is no requirement for any of the rooms to be of have a `type` of
`m.space`, any room with `m.space.child` events is considered.

Example request:

```jsonc
POST /_matrix/client/r0/rooms/{roomID}/spaces

{
    "max_rooms_per_space": 5,
    "suggested_only": true
}
```

or:

```text
GET /_matrix/client/r0/rooms/{roomID}/spaces?
    max_rooms_per_space=5&
    suggested_only=true
```

> Example response:

Reviewer 9: Somewhere between the request and response it would be nice to say the words "The endpoint requires authentication and can be rate limited". Or, if these aren't conditions on the endpoint, explanations as to why not.

```jsonc
{
    "rooms": [
        {
            "room_id": "!ol19s:bleecker.street",
            "avatar_url": "mxc://bleecker.street/CHEDDARandBRIE",
            "guest_can_join": false,
            "name": "CHEESE",
            "num_joined_members": 37,
            "topic": "Tasty tasty cheese",
            "world_readable": true,
            "creation_ts": 1432735824653,
            "room_type": "m.space"
        },
        { ... }
    ],
    "events": [
        {
            "type": "m.space.child",
            "state_key": "!efgh:example.com",
            "content": {
                "via": ["example.com"],
                "suggested": true
            },
            "room_id": "!ol19s:bleecker.street",
            "sender": "@alice:bleecker.street"
        },
        { ... }
    ]
}
```

Request params:

* **`suggested_only`**: Optional. If `true`, return only child events and rooms where the
  `m.space.child` event has `suggested: true`. Defaults to `false`.

  For the POST request, must be a boolean. For GET, must be either `true` or `false`,

>   case-sensitive.

Reviewer 9: as far as the spec is concerned, both of these cases are booleans. We don't get into the details of query strings technically being string pairs.

Reviewer 8: I'm not sure I understand, are you saying to just remove this paragraph?

Reviewer 9: just shortening it to "must be a boolean" would be fine, probably merging it into some other text somewhere.

* **`max_rooms_per_space`**: Optional: a client-defined limit to the maximum
  number of children to return per space. Doesn't apply to the root space (ie,
  the `room_id` in the request).

  Server implementations may also have an internal limit (recommended as 50)

>   (which *does* apply to the root room); attempts to exceed this  limit are

Reviewer 9: why is the server's internal limit different than the provided limit? Seems like it'd just be even more complicated for servers than `min(provided, 50)`

Reviewer 8: This paragraph is pretty much saying use `min(provided, 50)`, but 50 is just a recommendation.

Reviewer 9: I'd suggest leaving it at "the server might impose a more restrictive limit than the client requested" - the details aren't important for the MSC/spec.

  ignored. Must be a non-negative integer.

Response fields:

* **`rooms`**: for each room/space, starting with the root room, a
  summary of that room. The fields are the same as those returned by
  `/publicRooms` (see
  [spec](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-client-r0-publicrooms)),
  with the addition of:
  * **`room_type`**: the value of the `m.type` field from the
    room's `m.room.create` event, if any.
  * **`creation_ts`**: the value of the `origin_server_ts` field from the
    room's `m.room.create` event. This is required for sorting of rooms as specified
    in [MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772).

> * **`events`**: `m.space.child` events of the returned rooms. For each event, only the

Reviewer 9: it might be best to call this `stripped_state`, `space_state`, or similar to line up with terminology already established in the spec for this data structure. Specifically, these look like StrippedState events with the addition of a room_id

Reviewer 8: As discussed below these are stripped state events. We can explore renaming the field.

> * **`events`**: `m.space.child` events of the returned rooms. For each event, only the

Reviewer 9: a few words as to why this is not under the relevant room would be appreciated.

Reviewer 8: I'm not sure there's a particular reason, I believe the child events could be added under the room.

Reviewer 8: @Kegsay / @richvdh Any idea why these were not added under the rooms themselves? I'm guessing this was due to the original implementation going up and down the tree or something?

Reviewer 6: @Kegsay might know better, but I think the idea is that it's possible under certain circumstances - particularly with pagination? - to return events without their parent rooms.

Reviewer 2: > I'm guessing this was due to the original implementation going up and down the tree or something?

This is the main reason. The previous impl had children and parents, so it wasn't clear where each event should live. The new impl just has children which are events in the room, so it could in theory live under there. However, this wouldn't play nicely with any kind of pagination as vdh points out: if you have a room (so you know the name, topic, etc) you don't want to have to return that data _again_ when returning the next page of results for the room.

  following fields are returned: `type`, `state_key`, `content`, `room_id`,

>   `sender`.

Reviewer 9: why not the whole event? do we need to worry about event IDs here?

Reviewer 8: As you mentioned above, this is pretty much the stripped state, with the room ID (so you can map it to the rooms above).

Reviewer 9: right, but why is this stripped state instead of the whole event? It's not immediately clear in the MSC why we're intentionally using stripped state over the client-server formatted events.

Reviewer 8: @Kegsay / @richvdh Not sure why this was chosen in particular. Was it just because it felt that the entire event was not needed and this reduces API size? Is it due to not wanting to leak more information than necessary?

Reviewer 2: For a few reasons:
 - To make sure we leak as little metadata as possible (think timestamps, `prev_content`, etc)
 - To make sure clients treat this data separately to normal events. Much like with invite state, the data here may not be trustworthy (if it came over federation) and is no substitute to the actual events.

Reviewer 9: fair, would be good to document those details in the MSC.

Reviewer 8: Thanks so much for the context. I've added this to the MSC.

  
Errors:

403 with an error code of `M_FORBIDDEN`: if the user doesn't have permission to

> view/peek the root room (including if  that room does not exist).

Reviewer 9: for not found, shouldn't that be M_NOT_FOUND?

Reviewer 8: This matches the behavior of other room endpoints (e.g. [`/aliases`](https://matrix.org/docs/spec/client_server/latest#get-matrix-client-r0-rooms-roomid-aliases)). I assume it is to not divulge whether the user doesn't have permission or the room doesn't exist.

Reviewer 9: ah, would be good to get that into the MSC then. That makes sense :)

5. Recurse into any newly added targets of `m.space.child` events (i.e. repeat
   step 4), until either all discovered rooms have been  inspected, or the
   server-side limit on the number of rooms is reached.

Other notes:

* Any inaccessible children are omitted from the result, but the `m.space.child`
  events that point to them are still returned.
* There could be loops in the returned child events - clients should handle this
  gracefully.
* Similarly, note that a child room might appear multiple times (e.g. also be a
  grandchild). Clients and servers should handle this appropriately.
* `suggested_only` applies transitively.

  For example, if a space A has child space B which is *not* suggested, and space
  B has suggested child room C, and the client makes a summary request for A with
  `suggested_only=true`,  neither B **nor** C will be returned.

  Similarly, if a space A has child space B which is suggested, and space B has
  suggested child room C which is suggested, and the client makes a summary request
  for A with `suggested_only=true`, both B and C will be returned.
* The current implementation doesn't honour `order` fields in child events, as

>   suggested in [MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772).

Reviewer 9: is it supposed to? Implementation notes in the proposal section are a bit unexpected.

Reviewer 8: Ordering the results can potentially be really hard due to not necessarily having state of all the children of a room. If you don't have state you don't have the `origin_server_ts` of the `m.room.create` even, which is required for ordering according to MSC1772.

I think it might be an improvement to modify the ordering described in MSC1772 to only use data available in the children events instead of assuming the homeserver is in all of those rooms. I'd be curious on the thoughts of others here though! (I think that using the `origin_server_ts` was essentially an arbitrary tie-breaker anyway, so removing it shouldn't be too controversial.)

Reviewer 8: matrix-org/synapse#9954 implements this, and c142433c65ff02e799575b1989a88660c8bddf3b adds some notes about this to the MSC itself.

* `m.space.child` with an invalid `via` (invalid is defined as missing, not an
  array or an empty array) are ignored.

### Server-server API

Much the same interface as the Client-Server API.

Example request:

```jsonc

> POST /_matrix/federation/v1/spaces/{roomID}

Reviewer 9: POST/GET argument extends to this as well.

Reviewer 8: I think I'd rather get rid of the both forms for the C-S API than to add them both to the S-S API. 🤷

Reviewer 6: agreed, we should just spec one. It's a read-only API so it should be GET imho.

Reviewer 8: I've updated both of these to be `GET` only!

    "suggested_only": true
}
```

The response has the same shape as the Client-Server API.

Request params are the same as the Client-Server API, with the addition of:

* **`exclude_rooms`**: Optional. A list of room IDs that can be omitted
  from the response.

This is largely the same as the Client-Server API, but differences are:

* The calling server can specify a list of spaces/rooms to omit from the
  response (via `exclude_rooms`).
* `max_rooms_per_space` applies to the root room as well as any returned
  children.
* If the target server is not a member of any discovered children (so
  would have to send another request over federation to inspect them), no
  attempt is made to recurse into them - they are simply omitted from the
  response.
  * If the target server is not a member of the root room, an empty
    response is returned.
* Currently, no consideration is given to room membership: the spaces/rooms
  must be world-readable (ie, peekable) for them to appear in the results.

## Potential issues

To reduce complexity, only a limited number of rooms are returned for a room,
no effort is made to paginate the results. Proper pagination is left to a future

> MSC.

Reviewer 9: Pagination seems pretty important for a feature like this: what is the motivation for excluding it?

Reviewer 8: I believe the conversation for not including pagination was that it is extremely complicated and unclear how much benefit it would bring (are cases of having over 50 rooms in a space common?), but maybe @richvdh or @t3chguy can elaborate as they were more involved in that conversation.

Reviewer 4: Pagination is definitely desired. @richvdh might be better to comment on this but the concern was that we don't know what the shape of the pagination mechanism should be, it could be that this API is breadth first up to a limit but always returning at least an entire depth, then you could just call this same API on a different root node as a form of traversal, or whether it should be more like the room directory/room pagination with an opaque token.

Reviewer 8: I've thought about this a little bit and I *think* that the idea of traversing by providing different root nodes could lead to many request/response cycles in the case of a deep and narrow hierarchy. You might be able to pair this with greedily continuing down layers until a limit is reached, but that sounds complicated. This also seems difficult to guard against loops, but maybe that could be left up to the calling client?

In the opposite case (a shallow, wide hierarchy) you run into the issue where you might need to paginate the current depth, but this seems straightforward (you just need the node whose children are being fetched and the last returned result).

I think something similar to the current API, but including a pagination token might work. The server would need to keep track of quite a bit of state to continue the processing though, including: the processed rooms, the processed events, and the pending queue of rooms to look at. You could probably serialize those into a large pagination token, or store it in the database.

Overall, I think traversing via the tree (by choosing different roots) feels more elegant?

(assuming they have permission to) but this is costly and slow.

### Client API

```
POST /_matrix/client/r0/rooms/{roomID}/spaces
{
    "max_rooms_per_space": 5,      // The maximum number of rooms/subspaces to return for a given space, if negative unbounded. default: -1.
    "auto_join_only": true,        // If true, only return m.space.child events with auto_join:true, default: false, which returns all events.
    "limit": 100,                  // The maximum number of rooms/subspaces to return, server can override this, default: 100.
    "batch": "opaque_string"       // A token to use if this is a subsequent HTTP hit, default: "".
}
```

which returns:

```
{
    "next_batch": "opaque string",
    "rooms": [
        {
            "aliases": [
                "#murrays:cheese.bar"
            ],
            "avatar_url": "mxc://bleeker.street/CHEDDARandBRIE",
            "guest_can_join": false,
            "name": "CHEESE",
            "num_joined_members": 37,
            "room_id": "!ol19s:bleecker.street",
            "topic": "Tasty tasty cheese",
            "world_readable": true,

            "num_refs": 42,

>             "room_type": "m.space"

Reviewer 4: Presumably this will have a `canonical_alias` field too if available

Reviewer 2: Yes, the `PublicRoomsChunk` in the spec includes:
```
canonical_alias | string | The canonical alias of the room, if any.
```

  - This guarantees that all edges for the root node are given to the client. Not all edges of subspaces
    will be returned, nor will edges of all rooms be returned. This can be detected by clients in two ways:
      * Comparing `num_refs` with the *total number* of edges pointing to/from the room.
      * Comparing the number of `m.space.child` state events in the room with `max_rooms_per_space`, where
        `max_rooms_per_space` is 1 greater than the actual desired maximum value.
  - If not all events were returned due to reaching a `limit` or `max_rooms_per_space`, return a
    `next_batch` token. The server SHOULD NOT return duplicate events or rooms on subsequent
    requests: this can be achieved by remembering the event/room IDs returned to the caller between calls.
    This results in each request uncovering more nodes/edges until the entire tree has been explored.

Client behaviour:
 - Decide which room should be the root of the tree, then call this endpoint with the root room ID.
 - The data in `rooms` determines _what_ to show. The events in `events` determine _where_ to show it.
   Take all the data in `rooms` and key them by room ID.

>  - Loop through the `events` and keep track of parent->child relationships by looking at the `room_id`

Reviewer 4: in the example above `room_id` appears to be missing

## Spaces Summary API

*This MSC depends on [MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772).*

Spaces are rooms with `m.space` as the [room type](https://github.com/matrix-org/matrix-doc/pull/1840).
Spaces can include state events to specify parent/child relationships.
These relationships point to other rooms, which may themselves be spaces.
This means spaces can have subspaces and rooms. This creates a graph: a space directory.

This MSC defines a new endpoint which can be used to reveal information about the space directory.

Consider the graph:
```
    A
    ^
    |___
    |   |
    V   V
    B   R1
    ^
    |
    V
    R2

R1,R2 = rooms
A,B   = spaces
<-->  = parent/child relationship events
```
This MSC aims to create a way for clients to produce a tree view along the lines of:
```
Space A
   |
   |___ Room 1
   |
   Space B
     |
     |___ Room 2
```
Clients are able to do this currently by peeking into all of these rooms
(assuming they have permission to) but this is costly and slow.

### Client API

```

> POST /_matrix/client/r0/rooms/{roomID}/spaces

Reviewer 4: /me proposes `/_matrix/client/unstable/org.matrix.msc2946`

{
    "max_rooms_per_space": 5,      // The maximum number of rooms/subspaces to return for a given space, if negative unbounded. default: -1.
    "auto_join_only": true,        // If true, only return m.space.child events with auto_join:true, default: false, which returns all events.
    "limit": 100,                  // The maximum number of rooms/subspaces to return, server can override this, default: 100.
    "batch": "opaque_string"       // A token to use if this is a subsequent HTTP hit, default: "".
}
```

which returns:

```
{
    "next_batch": "opaque string",
    "rooms": [
        {
            "aliases": [
                "#murrays:cheese.bar"
            ],
            "avatar_url": "mxc://bleeker.street/CHEDDARandBRIE",
            "guest_can_join": false,
            "name": "CHEESE",
            "num_joined_members": 37,
            "room_id": "!ol19s:bleecker.street",
            "topic": "Tasty tasty cheese",
            "world_readable": true,

>             "num_refs": 42,

Reviewer 6: I'm going to unresolve this, because it would really be *WAY* easier to read this if the fields were documented more obviously rather than in the long "Justifications" text.

        this room as a parent.
     * `room_type` which is the room type, which is `m.space` for subspaces. It can be omitted if there is no room type
       in which case it should be interpreted as a normal room.
 - `events`: These are the edges of the graph. The objects in the array are stripped `m.space.parent`
   or `m.space.child` events. This means that they only contain the `type`, `state_key`, `content`, `room_id` and `sender`
   keys, similar to `invite_state` in the `/sync` API.
 - `next_batch`: Its presence indicates that there are more results to return.

Server behaviour:
 - Extract the room ID from the request. Sanity check request data. Begin walking the graph
   starting with the room ID in the request in a queue of unvisited rooms according to the
   following rules:
    * If this room has already been processed, skip. NB: do not remember this between calls,
      as servers will need to visit the same room more than once to return additional events.
    * Mark this room as processed.
    * Is the caller currently joined to the room or is the room `world_readable`?
      If no, skip this room. If yes, continue.

>     * If this room has not ever been in `rooms` (across multiple requests), extract the

Reviewer 2: The intention is to deliver each `Room` object exactly once to the client hence "has not ever been in `rooms`". However, when walking the tree, you will need to visit the node multiple times hence "if this room has been processed, skip".

## Spaces Summary API

> *This MSC depends on [MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772).*

Reviewer 6: actually sorry, that's not really right. It *does* depend on significant parts of MSC1772. I wish MSC1772 wasn't so sprawling.

```suggestion
*This MSC depends on parts of [MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772).*
```

Spaces are rooms with `m.space` as the [room type](https://github.com/matrix-org/matrix-doc/pull/1840).
Spaces can include state events to specify parent/child relationships.
These relationships point to other rooms, which may themselves be spaces.
This means spaces can have subspaces and rooms. This creates a graph: a space directory.

This MSC defines a new endpoint which can be used to reveal information about the space directory.

Consider the graph:
```
    A
    ^
    |___
    |   |
    V   V
    B   R1
    ^
    |
    V
    R2

R1,R2 = rooms
A,B   = spaces
<-->  = parent/child relationship events
```
This MSC aims to create a way for clients to produce a tree view along the lines of:
```
Space A
   |
   |___ Room 1
   |
   Space B
     |
     |___ Room 2
```
Clients are able to do this currently by peeking into all of these rooms
(assuming they have permission to) but this is costly and slow.

### Client API

```

> POST /_matrix/client/r0/rooms/{roomID}/spaces

Reviewer 6: should guest users be allowed to call this API? I assume so?

Reviewer 4: Sounds like it could be a potentially expensive API call

Reviewer 4: Guests don't need to pass captcha which makes this infinitely easier to abuse

Reviewer 2: Aside from the computational cost, there's no reason why guests should be prevented from hitting this endpoint.

Reviewer 10: Seems like abuse should be countered with rate limiting or other things not just excluding guests from this endpoint. A guest-supporting client might want to preview a whole space to the user before deciding to join / register etc.

Reviewer 4: Excessive rate limiting could harm the usability of this for non-abusive users just switching between spaces a few times to find the room they want

Reviewer 9: This feels like it could be heavily cached by the server and ultimately become a public room directory-like endpoint instead.

> POST /_matrix/client/r0/rooms/{roomID}/spaces

Reviewer 6: why is this a POST rather than a GET?

Reviewer 2: There's a decent number of input parameters and that only looks set to grow (i.e suggested rooms, auto join rooms, etc). If it's too big it might cause problems being query params.

Reviewer 9: It seems like since this comment thread happened there spawned a GET version in addition to the POST version: why is there still a POST version?

{

>     "max_rooms_per_space": 5,      // The maximum number of rooms/subspaces to return for a given space, if negative unbounded. default: -1.

Reviewer 6: why negative to mean unbounded, rather than simply omitting the param? we don't do this anywhere else in the API afaik.

Reviewer 2: Sounds good to me.

>     "max_rooms_per_space": 5,      // The maximum number of rooms/subspaces to return for a given space, if negative unbounded. default: -1.

Reviewer 6: it's not made clear here, but apparently the root space is exempted from this limit.

>     "max_rooms_per_space": 5,      // The maximum number of rooms/subspaces to return for a given space, if negative unbounded. default: -1.

Reviewer 6: Another thing that is not made clear here: AIUI this limit applies to both edge events, and the rooms they point to. 

Example: if a space A has 10 child events, and `max_rooms_per_space` is set to 5:
 * for the first 5 events, we add the event to `events`, and recurse into the rooms they point to (assuming we haven't already visited them)
 * we completely ignore any additional events.

Reviewer 2: Yes, that's correct.

    "auto_join_only": true,        // If true, only return m.space.child events with auto_join:true, default: false, which returns all events.

>     "limit": 100,                  // The maximum number of rooms/subspaces to return, server can override this, default: 100.

Reviewer 8: ```suggestion
    "limit": 100,                  // The maximum number of events to return, server can override this, default: 100.
```

Maybe? I think this matches what is described below. (I was quite confused by the current wording of `max_rooms_per_space` vs. `limit`.)

Reviewer 6: you are very much not alone in your confusion...

    "batch": "opaque_string"       // A token to use if this is a subsequent HTTP hit, default: "".
}
```

which returns:

```
{
    "next_batch": "opaque string",
    "rooms": [
        {
            "aliases": [
                "#murrays:cheese.bar"

>             ],

Reviewer 6: as an aside: note that `aliases` is no longer returned by synapse here (https://github.com/matrix-org/matrix-doc/issues/3056).

            "avatar_url": "mxc://bleeker.street/CHEDDARandBRIE",
            "guest_can_join": false,
            "name": "CHEESE",
            "num_joined_members": 37,
            "room_id": "!ol19s:bleecker.street",
            "topic": "Tasty tasty cheese",
            "world_readable": true,

            "num_refs": 42,
            "room_type": "m.space"
        },
        { ... }
    ],

>     "events": [

Reviewer 6: according to the description, seems it does.

            },
            "room_id": "!ol19s:bleecker.street",
            "sender": "@alice:bleecker.street"
        }
    ]
}
```

Justifications for the request API shape are as follows:
 - The HTTP path: Spaces are scoped to a specific room to act as an anchor point for
   navigating the directory. Alternatives are `/r0/spaces` with `room_id` inside the
   body, but this feels less idiomatic for room-scoped requests.
 - The HTTP method: there's a lot of data to provide to the server, and GET requests
   shouldn't have an HTTP body, hence opting for POST. The same request can produce
   different results over time so PUT isn't acceptable as an alternative.
 - `max_rooms_per_space`: UIs can only display a set number of rooms per space, so allowing
   clients to specify this limit is desirable. Subsequent rooms can be obtained by paginating.
   The graph has 2 distinct types of nodes, and some UIs may want to weight one type above
   the other. However, it's impossible to always know what type of node a given room ID falls
   under because the server may not be joined to that room (to determine the room type) or the
   caller may not have permission to see this information.

>  - `limit`: The maximum number of events to return in `events`. It is desirable for clients

Reviewer 6: should this have a name which makes it clear it applies to the number of `events`, not the number of rooms?

Reviewer 2: That would be useful yeah, possibly `event_limit`?

   and servers to be able to put a maximum cap on the amount of data returned to the client.
   **This limit may be exceeded if the root room has `> limit` rooms.**
 - `auto_join_only`: If `true`, only a subset of the graph is returned based on the presence
   of `auto_join: true` in the `content` field of `m.space.child`. Some clients may only
   care about the "main" or "default" rooms, which are rooms with this flag set. This does
   not affect parent state events: they are still returned. This does not modify the value
   of `num_refs`.

>  - `batch`: Required for pagination. Could be a query parameter but it's easier if

Reviewer 6: why is this called `batch`? we don't seem to use that name anywhere else. Alternatives might include `from` (for consistency with `/messages`) or `since` (for consistency with `/publicRooms`).

Reviewer 2: Mainly because `/sync` has `next_batch` hence why I went with that term. I don't mind `since`.

 - `rooms`: These are the nodes of the graph. The objects in the array are exactly the same as `PublicRoomsChunk` in the
   [specification](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-client-r0-publicrooms)
   as the information displayed to users is the same. There are two _additional_ keys
   which are:
     * `num_refs` which is the total number of state events which point to or from this room (inbound/outbound edges).
        This includes all `m.space.child` events in the room, _in addition to_ `m.space.parent` events which point to
        this room as a parent.
     * `room_type` which is the room type, which is `m.space` for subspaces. It can be omitted if there is no room type
       in which case it should be interpreted as a normal room.
 - `events`: These are the edges of the graph. The objects in the array are stripped `m.space.parent`
   or `m.space.child` events. This means that they only contain the `type`, `state_key`, `content`, `room_id` and `sender`
   keys, similar to `invite_state` in the `/sync` API.
 - `next_batch`: Its presence indicates that there are more results to return.

Server behaviour:

>  - Extract the room ID from the request. Sanity check request data. Begin walking the graph

Reviewer 6: what should the server do if the supplied room id is not a spaceroom?

Reviewer 2: This API doesn't look at the create event to see if the room type is a space, it only looks at state events. This feels somewhat desirable, though unexpected. We should probably reject at this stage if the room type isn't a space.

   starting with the room ID in the request in a queue of unvisited rooms according to the
   following rules:
    * If this room has already been processed, skip. NB: do not remember this between calls,
      as servers will need to visit the same room more than once to return additional events.
    * Mark this room as processed.
    * Is the caller currently joined to the room or is the room `world_readable`?

>       If no, skip this room. If yes, continue.

Reviewer 6: if the room is not accessible, should we continue into its children/parents or not?

Reviewer 2: No, if the room is not accessible children/parents should not be explored.

      to the caller.
    * If this is the root room from the original request, insert all these events into `events` if
      they haven't been added before (across multiple requests).
    * Else add them to `events` honouring the `limit` and `max_rooms_per_space` values. If either
      are exceeded, stop adding events. If the event has already been added, do not add it again.
    * For each referenced room ID in the events being returned to the caller (both parent and child)
      add the room ID to the queue of unvisited rooms. Loop from the beginning.
  - This guarantees that all edges for the root node are given to the client. Not all edges of subspaces
    will be returned, nor will edges of all rooms be returned. This can be detected by clients in two ways:
      * Comparing `num_refs` with the *total number* of edges pointing to/from the room.
      * Comparing the number of `m.space.child` state events in the room with `max_rooms_per_space`, where
        `max_rooms_per_space` is 1 greater than the actual desired maximum value.
  - If not all events were returned due to reaching a `limit` or `max_rooms_per_space`, return a

>     `next_batch` token. The server SHOULD NOT return duplicate events or rooms on subsequent

Reviewer 6: I'm a bit confused about how the different limits and the pagination parameters interact. If we have a space with 1000 children (so we hit the max_rooms_per_space), does paginating work its way through all of those edges in preference to moving onto other spaces in the hierarchy?

Reviewer 2: The intention behind `max_rooms_per_space` is to allow hierarchies like:
```
Root
|
|_A
| |_ Room 1
| |_ Room 2
|
|_B
  |_ Room 3
  |_ Room 4
  |_ Room 5
  |_ ... [hundreds]
```

to be roughly obtained in a single API call. That is to say, a single large subspace shouldn't starve neighbouring spaces, hence having a "per room" (although per space would be better terminology) limit. The intention is that if you then wanted to find more rooms in B, you would do this as a separate API call (e.g hitting a "And `$num_refs` more..." button). How pagination of this sub-space interacts with pagination of the root space I don't fully know the answer to. I said before that we _should_ still be returning sub-space results, but it would probably be more useful to not, and instead expect a separate API call with B as the root.

   children.
 - Starting at the root room ID:
     * Compare the `num_refs` value in `rooms.$room_id` to the total number of events which reference this
       room in `events` (across all rooms). If they differ, a partial response has been returned for this
       space and additional results should be loaded when required. The API guarantees that *all* events for
       the root room ID will be returned, regardless of how many events there are (even if they exceed `limit`).
     * Lookup all children for this room ID. For each child:
        - If there is no corresponding room data for this room ID then this room is either a subspace or a room.
          The room is not world readable or the server does not have any information about this room. Clients
          MAY be able to join this room by issuing a `/join` request.
        - If the child is a room (not a space, check the `room_type` field), look up the room data from
          `rooms` and render it.
        - Else the child is a space, render the space as a heading (using the room name/topic) and
          restart the lookup using the new space room ID.

> ### Federation API

Reviewer 6: note to self: we need to spec the "I'm not in this space" behaviour, so that the calling server can try a different server.

  with the addition of:
  * **`room_type`**: the value of the `m.type` field from the
    room's `m.room.create` event, if any.
* **`events`**: `m.space.child` events of the returned rooms. For each event, only the
  following fields are returned: `type`, `state_key`, `content`, `room_id`,
  `sender`.<sup id="a1">[1](#f1)</sup>

Errors:

403 with an error code of `M_FORBIDDEN`: if the user doesn't have permission to
view/peek the root room (including if that room does not exist). This matches the
behavior of other room endpoints (e.g.
[`/_matrix/client/r0/rooms/{roomID}/aliases`](https://matrix.org/docs/spec/client_server/latest#get-matrix-client-r0-rooms-roomid-aliases)).
To not divulge whether the user doesn't have permission vs whether the room
does not exist.

#### Algorithm

A rough algorithm follows:

1. Start at the "root" room (the provided room ID).
2. Generate a summary and add it to `rooms`.
3. Add any `m.space.child` events in the room to `events`.
4. Recurse into the targets of the `m.space.child` events.
   1. If the room is not accessible (or has already been processed), do not
      process it.
   2. Generate a summary for the room and add it to `rooms`.

>    3. Add any `m.space.child` events of the room to `events`.

Reviewer 8: This causes a bit of an odd thing to happen (see matrix-org/synapse#10020) where you can end up with an event that mentions a room, but then no room entry. It is unclear *why* you don't have the room entry (it could be either due to reaching the limit of the API response *or* due to not having access to the room).

Is having the event here useful? What purpose does it serve?

(Note that this is not any additional metadata leak -- the event is part of the state of the room so is accessible to the requester anyway.)

Reviewer 8: I've clarified the bit below about when these links to nowhere are turned.

5. Recurse into any newly added targets of `m.space.child` events (i.e. repeat
   step 4), until either all discovered rooms have been inspected, or the
   server-side limit on the number of rooms is reached.

   In the case of the homeserver not having access to the state of a room, the
   server-server API (see below) can be used to query for this information over
   federation from one of the servers provided in the `via` key of the
   `m.space.child` event.

Other notes:

* Any inaccessible children are omitted from the result, but the `m.space.child`

>   events that point to them are still returned.

Reviewer 8: We should probably define more clearly what "accessible" means here.

Reviewer 8: This was done in 3b0051fb103f150810901375a2191ae9c1b2e566.

`2` it cannot fulfill this request since it is unsure how to order Room B, Room
C, and Room D, but it can only return 2 of them. It *can* reach out over
federation to HS2 and request a space summary for Room D, but this is undesirable:

* HS1 might not have the permissions to know any  of the state of Room D, so might
  receive a 403 error.
* If we expand the example above to many rooms than this becomes expensive to
  query a remote server simply for ordering.

This proposes changing the ordering rules from MSC1772 to the following:

* Rooms are sorted based on a lexicographic ordering of the Unicode codepoints
  of the characters in `order` values.

  `order`s which are not strings, or do not consist solely  of ascii characters
  in the range `\x20` (space) to `\x7F` (~), or consist of more than 50
  characters, are forbidden and the field should be ignored if received.
* Rooms with no `order` come last, in ascending lexicographic order of their
  `room_id`s.

This removes the clauses discussing using the `origin_server_ts` of the
`m.room.create` event to allow a defined sorting of siblings based purely on the

> information available in the `m.space.child` event.

Reviewer 4: The point of this was to make newer/older sub-spaces/rooms appear last naturally, using roomId won't achieve this and misses the goal.
How about using the ts of the relationship event?

Reviewer 8: >  How about using the ts of the relationship event?

This would be fine with me! I've updated this with the above info!

>
> We refer to such collections of rooms as "spaces".

This MSC attempts to solve how a member of a space discovers rooms in that space. This
is useful for quickly exposing a user to many aspects of an entire community, using the
examples above, joining the "official matrix.org rooms" space might suggest joining a few
rooms:

* A room to discuss development of the Matrix Spec.
* An announcements room for news related to matrix.org.
* An off-topic room for members of the space.

## Proposal

A new client-server API (and corresponding server-server API) is added which allows
for querying for the rooms and spaces contained within a space. This allows a client
to efficiently display a hierarchy of rooms to a user (i.e. without having
to walk the full state of the space).

### Client-server API

An endpoint is provided to walk the space tree, starting at the provided room ID
("the root room"), and visiting other rooms/spaces found via `m.space.child`
events. It recurses into the children and into their children, etc.

Note that there is no requirement for any of the rooms to be of have a `type` of

> `m.space`, any room with `m.space.child` events is considered.

Reviewer 8: I'm not sure why we do this -- it seems that at least Element Web strips out these results if it isn't a space, so why are we sending them? Why would we not consider the type?

Reviewer 8: I've reworked this to only include space rooms.

* **`events`**: `m.space.child` events of the returned rooms. For each event, only the
  following fields are returned: `type`, `state_key`, `content`, `room_id`,
  `sender`, <sup id="a1">[1](#f1)</sup> with the addition of:
  * **`creation_ts`**: the value of the `origin_server_ts` field from the
    `m.space.child` event. This is required for sorting of rooms as specified
    in [MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772) and updated
    below.

Errors:

403 with an error code of `M_FORBIDDEN`: if the user doesn't have permission to
view/peek the root room (including if that room does not exist). This matches the
behavior of other room endpoints (e.g.
[`/_matrix/client/r0/rooms/{roomID}/aliases`](https://matrix.org/docs/spec/client_server/latest#get-matrix-client-r0-rooms-roomid-aliases)).
To not divulge whether the user doesn't have permission vs whether the room
does not exist.

#### Algorithm

A rough algorithm follows:

1. Start at the "root" room (the provided room ID).
2. Generate a summary and add it to `rooms`.
3. Add any `m.space.child` events in the room to `events`.
4. Recurse into the targets of the `m.space.child` events.

>    1. If the room is inaccessible (as defined by [room history visibility](https://matrix.org/docs/spec/client_server/latest#id87))

Reviewer 8: We should probably handle something about knocking (once that's implemented) to allow rooms to be visible if they're knockable. Or maybe we should define "accessible" as "able to join" (see [MSC3173](https://github.com/matrix-org/matrix-doc/pull/3173)).

  should handle this  gracefully.
* Similarly, note that a child room might appear multiple times (e.g. also be a
  grandchild). Clients and servers should handle this appropriately.
* `suggested_only` applies transitively.

  For example, if a space A has child space B which is *not* suggested, and space
  B has suggested child room C, and the client makes a summary request for A with
  `suggested_only=true`, neither B **nor** C will be returned.

  Similarly, if a space A has child space B which is suggested, and space B has
  suggested child room C which is suggested, and the client makes a summary request
  for A with `suggested_only=true`, both B and C will be returned.
* `m.space.child` with an invalid `via` (invalid is defined as missing, not an
  array or an empty array) are ignored.

### Server-server API

The Server-Server API has almost the same interface as the Client-Server API.
It is used when a homeserver does not have the state of a room to include in the
summary.

Example request:

```jsonc
GET /_matrix/federation/v1/spaces/{roomID}?
    exclude_rooms=%21a%3Ab&
    exclude_rooms=%21b%3Ac&
    max_rooms_per_space=5&
    suggested_only=true&
```

> The response has the same shape as the Client-Server API.

Reviewer 8: This needs updates based on changes to MSC3083, see https://github.com/matrix-org/matrix-doc/pull/3083/commits/0f4961118eddaf2959ad1d765dd3a0c8422df0a6

for querying for the rooms and spaces contained within a space. This allows a client
to efficiently display a hierarchy of rooms to a user (i.e. without having
to walk the full state of the space).

### Client-server API

An endpoint is provided to walk the space tree, starting at the provided room ID
("the root room"), and visiting other rooms/spaces found via `m.space.child`
events. It recurses into the children and into their children, etc.

Note that only rooms that have a `type` of `m.space` are considered when searching
for `m.space.child` events.

This endpoint requires authentication and is not subject to rate-limiting.

Example request:

```text
GET /_matrix/client/r0/rooms/{roomID}/spaces?
    max_rooms_per_space=5&
    suggested_only=true
```

Example response:

```jsonc
{
    "rooms": [
        {
            "room_id": "!ol19s:bleecker.street",
            "avatar_url": "mxc://bleecker.street/CHEDDARandBRIE",
            "guest_can_join": false,
            "name": "CHEESE",
            "num_joined_members": 37,
            "topic": "Tasty tasty cheese",
            "world_readable": true,
            "join_rules": "public",
            "room_type": "m.space",

>             "allowed_spaces": ["!abcdef:bleecker.street"]

Reviewer 4: This field got renamed but also should not be present in the cs api response, only in the ss api

  Similarly, if a space A has child space B which is suggested, and space B has
  suggested child room C which is suggested, and the client makes a summary request
  for A with `suggested_only=true`, both B and C will be returned.
* `m.space.child` with an invalid `via` (invalid is defined as missing, not an
  array or an empty array) are ignored.

### Server-server API

The Server-Server API has almost the same interface as the Client-Server API.
It is used when a homeserver does not have the state of a room to include in the
summary.

Example request:

```jsonc
GET /_matrix/federation/v1/spaces/{roomID}?
    exclude_rooms=%21a%3Ab&
    exclude_rooms=%21b%3Ac&
    max_rooms_per_space=5&
    suggested_only=true&
```

The response has the same shape as the Client-Server API.

Request parameters are the same as the Client-Server API, with the addition of:

* **`exclude_rooms`**: Optional. A list of room IDs that can be omitted
  from the response.

Response fields are the same as the Client-Server API, with the addition of:

* **`rooms`**: Each room contains an additional field:
  * **`allowed_room_ids`**: A list of room IDs which give access to this room per

>     [MSC3083](https://github.com/matrix-org/matrix-doc/pull/3083).

Reviewer 4: This got implemented as `allowed_spaces` - we need to either fix the MSC or the impl

  to no-limit. Must be a non-negative integer.

  Server implementations may wish to impose a maximum value to avoid resource
  exhaustion.
* **`from`**: Optional. Pagination token given to retrieve the next set of rooms.

#### Response Format

* **`rooms`**: `[object]` For each room/space, starting with the root room, a
  summary of that room. The fields are the same as those returned by
  `/publicRooms` (see
  [spec](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-client-r0-publicrooms)),
  with the addition of:
  * **`room_type`**: the value of the `m.type` field from the room's
    `m.room.create` event, if any.
  * **`children_state`**: The `m.space.child` events of the room. For each event,
    only the following fields are included<sup id="a1">[1](#f1)</sup>:
    `type`, `state_key`, `content`, `room_id`, `sender`,  with the addition of:
    * **`origin_server_ts`**: This is required for sorting of rooms as specified
      below.
* **`next_token`**: Optional `string`. The token to supply in the `from` param
  of the next `/spaces` request in order to request more rooms. If this is absent,

>   there are no more results.

Reviewer 6: With the notable exception of `/notifications`, `next_token` is otherwise unused in the Matrix API. I've attempted a survey of the various APIs which paginate in some way at https://github.com/matrix-org/matrix-doc/issues/1898; it's hard to discern a pattern but I think `next_batch` is marginally more common.

Reviewer 8: Ah, thanks for taking a look. I think I saw there was a variety and just picked one which sounded good.

Would you recommend using `next_batch` for both the response and request fields?

Reviewer 8: Also -- this currently states it is an error to change some parameters during pagination. I wonder if it makes more sense to just ignore those parameters (or perhaps make it an error to provide them at all with the pagination token)?

Reviewer 6: > Would you recommend using `next_batch` for both the response and request fields?

At this point I really have very little idea. Let me try and kick off a conversation about it in #1898.

Reviewer 8: I updated this to `from` / `next_batch`. We can change it again in the future if we want though.

            "children_state": [
                {
                    "type": "m.space.child",
                    "state_key": "!efgh:example.com",
                    "content": {
                        "via": ["example.com"],
                        "suggested": true
                    },
                    "room_id": "!ol19s:bleecker.street",
                    "sender": "@alice:bleecker.street",
                    "creation_ts": 1432735824653
                },
                { ... }
            ]
        },
        { ... }
    ],
    "next_token": "abcdef"
}
```

#### Errors:

An HTTP response with a status code of 403 and an error code of `M_FORBIDDEN`
should be returned if the user doesn't have permission to view/peek the root room.
This should also be returned if that room does not exist, which matches the
behavior of other room endpoints (e.g.
[`/_matrix/client/r0/rooms/{roomID}/aliases`](https://matrix.org/docs/spec/client_server/latest#get-matrix-client-r0-rooms-roomid-aliases))
to not divulge that a room exists which the user doesn't have permission to view.

An HTTP response with a status code of 400 and an error code of `M_INVALID_PARAM`

> should be returned if the `from` token provided is unknown to the server.

Reviewer 8: @t3chguy pointed out that we should go into a bit more depth about which (if any) parameters can be modified during pagination. E.g. changing `limit` seems very reasonable (e.g. pulling 50 items for a first page, then 10 at a time), while changing `max_depth` or `suggested_only` likely doesn't make sense.

to walk the full state of each room).

### Client-server API

An endpoint is provided to walk the space tree, starting at the provided room ID
("the root room"), and visiting other rooms/spaces found via `m.space.child`
events. It recurses into the children and into their children, etc.

Any child room that the user is joined or is potentially joinable (per
[MSC3173](https://github.com/matrix-org/matrix-doc/pull/3173)) is included in
the response. When a room with a `type` of `m.space` is found, it is searched
for valid `m.space.child` events to recurse into.

In order to provide a consistent experience, the space tree should be walked in
a depth-first manner, e.g. whenever a space is found it should be recursed into
by sorting the children rooms and iterating through them.

There could be loops in the returned child events; clients and servers should
handle this gracefully. Similarly, note that a child room might appear multiple
times (e.g. also be a grandchild). Clients and servers should handle this
appropriately.

This endpoint requires authentication and is not subject to rate-limiting.

#### Request format

```text

> GET /_matrix/client/r0/rooms/{roomID}/spaces

Reviewer 6: maybe `/children` ?

> GET /_matrix/client/r0/rooms/{roomID}/spaces

Reviewer 8: Some of the rooms will be spaces -- this returns the room you asked for, as well as children rooms (including spaces).

> GET /_matrix/client/r0/rooms/{roomID}/spaces

Reviewer 4: But I already implemented it as `/children` :D

https://github.com/matrix-org/matrix-js-sdk/pull/1818

> GET /_matrix/client/r0/rooms/{roomID}/spaces

Reviewer 8: (Although `/children` does seem a bit more functional. It returns the room itself too though, which is a bit odd.)

> GET /_matrix/client/r0/rooms/{roomID}/spaces

Reviewer 4: Good point. I'm not a huge fan of either the more I think about it. `/hierarchy` maybe?

> GET /_matrix/client/r0/rooms/{roomID}/spaces

Reviewer 6: > Some of the rooms will be spaces -- this returns the room you asked for, as well as children rooms (including spaces).

yes, some of them will be spaces, but many of them - most of them, even - will not be. It just felt a bit odd to be asking for `/spaces` and getting a load of results that weren't spaces. `/hierarchy` sounds sensible to me.

(to answer a point you made elsewhere - I realise this isn't new with this iteration of this API. It just felt a bit odd when I was re-reading it, and if you're changing other stuff anyway, we could revisit it.)

> GET /_matrix/client/r0/rooms/{roomID}/spaces

Reviewer 4: also changing the path to `/hierarchy` would mean we can keep the same prefix, not that it matters much at all

> GET /_matrix/client/r0/rooms/{roomID}/spaces

Reviewer 8: I'm OK with changing it to `/hierarchy` (and re-using the old prefix). I'll update it in a bit.

> GET /_matrix/client/r0/rooms/{roomID}/spaces

Reviewer 8: I've made this change.

  number of rooms to return per page. Must be a non-negative integer.

  Server implementations should impose a maximum value to avoid resource
  exhaustion.
* **`max_depth`**: Optional: The maximum depth in the tree (from the root room)
  to return. The deepest depth returned will not include children events. Defaults
  to no-limit. Must be a non-negative integer.

  Server implementations may wish to impose a maximum value to avoid resource
  exhaustion.
* **`from`**: Optional. Pagination token given to retrieve the next set of rooms.

#### Response Format

* **`rooms`**: `[object]` For each room/space, starting with the root room, a
  summary of that room. The fields are the same as those returned by
  `/publicRooms` (see
  [spec](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-client-r0-publicrooms)),
  with the addition of:
  * **`room_type`**: the value of the `m.type` field from the room's
    `m.room.create` event, if any.
  * **`children_state`**: The `m.space.child` events of the room. For each event,
    only the following fields are included<sup id="a1">[1](#f1)</sup>:
    `type`, `state_key`, `content`, `room_id`, `sender`,  with the addition of:

>     * **`creation_ts`**: the value of the `origin_server_ts` field from the

Reviewer 4: Sure but then you might think which `origin_server_ts` - the one of the latest event in the room? the earlier? neither, the one of the m.room.create event.

>     * **`creation_ts`**: the value of the `origin_server_ts` field from the

Reviewer 8: It is not from the create event, it is the one from the `m.space.child` event (this MSC attempts to change what MSC1772 says).

Pretty much all the fields are from the same event, so it is reasonable to use the "real" field names.

>     * **`creation_ts`**: the value of the `origin_server_ts` field from the

Reviewer 4: oh right, yes 🤦

# MSC2946: Spaces Summary

This MSC depends on [MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772), which
describes why a Space is useful:

> Collecting rooms together into groups is useful for a number of purposes. Examples include:
>
> * Allowing users to discover different rooms related to a particular topic: for example "official matrix.org rooms".
> * Allowing administrators to manage permissions across a number of rooms: for example "a new employee has joined my company and needs access to all of our rooms".
> * Letting users classify their rooms: for example, separating "work" from "personal" rooms.
>
> We refer to such collections of rooms as "spaces".

This MSC attempts to solve how a member of a space discovers rooms in that space. This
is useful for quickly exposing a user to many aspects of an entire community, using the
examples above, joining the "official matrix.org rooms" space might suggest joining a few
rooms:

* A room to discuss development of the Matrix Spec.
* An announcements room for news related to matrix.org.
* An off-topic room for members of the space.

> ## Proposal

Reviewer 9: the endpoint should exclude upgraded rooms where the server can see the new rooms.

Reviewer 4: and where that upgraded room is also a child of the same parent surely? otherwise rooms might just vanish if the user doing the upgrade isn't an admin in the space also (e.g personal spaces)

Reviewer 8: Meaning that the expectation is that when you upgrade a room you *don't* remove it from the space and just let both coexist, but only return the "latest" room or something like that?

Reviewer 9: yea, so the rooms don't get punted to Home if you haven't rejoined yet.

Reviewer 8: That seems reasonable on paper. I'll see how hard that is to implement.

Reviewer 8: I see a couple of potential issues with this:

* If the upgraded room is moved to a sub-space, would we want to check if it appears anywhere in the hierarchy? (I'd think not.) It could be confusing if you reorganize rooms that old tombstoned rooms reappear though.
* Would we return both rooms as part of the `children_state`? (I'd think so since they're both still there.)

Reviewer 6: > It could be confusing if you reorganize rooms that old tombstoned rooms reappear though.

This. To me it feels magical that we suppress the old room until the new room is removed from the space, at which point the old room suddenly reappears.

>     * Would we return both rooms as part of the `children_state`? (I'd think so since they're both still there.)

Probably? But won't it be odd that there are rooms referenced in `children_state` which we can't see?

Generally, I feel like this needs more thought, and implementation, and I don't really think we should block this MSC on it. I'd much rather we land what we have (and get it off unstable endpoints) despite a rough edge or two, than block it until we have good answers on the interaction of room upgrades with space summaries.

Reviewer 9: From what I understand of that conversation, the rough suggestion is that we'd expose a flag to show that a room/space is upgraded and likely where it points to (stripped state of the tombstone might work here, I think). This doesn't really solve the problem of the upgraded room moving into a subspace as part of the upgrade, though the general consensus appears to be that the old room is intended to be forgotten from the tree (at least in the UI), so the client could just trust that the new room being *anywhere* in the tree is enough to cause the old room to be hidden completely. Upgrading spaces still appears to have an issue where it could result in needless pagination requests, though room upgrades for (sub)spaces is generally an unsolved problem, so happy to punt that right out of this conversation (imo).

Of interesting note, https://github.com/matrix-org/matrix-react-sdk/pull/6906 solves much of the problem for when the client is aware of the upgraded room, though doesn't implement a local de-dupe approach from the upgraded flag mentioned above - this is easily something we can add later on.

Sorry for being moderately difficult on this: I really just want us to consider issues with implementation and spec before it causes problems in future MSCs. I realize that doing so drags out the MSC, which is by far not my favourite thing to do, but it's extremely important from a spec perspective that we don't fall into the trap of "well, it's released so let's approve the MSC".

> I'm really, really worried that we're very soon going to be in a position where migrating away from the unstable endpoints is just too difficult.

I honestly see this as a failure of the process, not of this thread. The SCT should have worked more closely with the teams/projects looking to ship the feature, and the teams/projects should have engaged the SCT to start working on a release date. In the end, no one appears to have engaged the other side (rightly so: it's not clear who should start this conversation), which ultimately means that spec review comes late when people feel as though the MSC can't change because the feature is in the wild and being relied upon. We've already talked about this problem at a spec retro, though that retro was too late for this MSC, and the conclusion is roughly "more communication". We should iron out the details of what that means exactly separate to all of this.

----

TLDR: Thank you for running the mental exercise, and sorry it's coming across as needlessly blocking. It's really not my intention to be an arbitrary pain/blocker for this MSC.

for querying for the rooms and spaces contained within a space. This allows a client
to efficiently display a hierarchy of rooms to a user (i.e. without having
to walk the full state of each room).

### Client-server API

An endpoint is provided to walk the space tree, starting at the provided room ID
("the root room"), and visiting other rooms/spaces found via `m.space.child`
events. It recurses into the children and into their children, etc.

Any child room that the user is joined or is potentially joinable (per
[MSC3173](https://github.com/matrix-org/matrix-doc/pull/3173)) is included in
the response. When a room with a `type` of `m.space` is found, it is searched
for valid `m.space.child` events to recurse into.

In order to provide a consistent experience, the space tree should be walked in
a depth-first manner, e.g. whenever a space is found it should be recursed into
by sorting the children rooms and iterating through them.

There could be loops in the returned child events; clients and servers should
handle this gracefully. Similarly, note that a child room might appear multiple
times (e.g. also be a grandchild). Clients and servers should handle this
appropriately.

> This endpoint requires authentication and is not subject to rate-limiting.

Reviewer 9: Rationale for why this endpoint is exempt from rate limiting would be appreciated.

Reviewer 8: I have no rationale, I arbitrarily picked not rate-limited and was hoping to have a conversation about it which was probably lost in the many comments above.

Reviewer 9: It's probably fair to consider this endpoint not rate limited for now, given client behaviour is quite tightly coupled to the endpoint. Servers can also cache this reasonably well on a per-room basis, and the response size is relatively fixed, making it easy to constrain resources on. 

Excessive abuse of the endpoint can (and should) lead to rate limiting, but that's not really what the spec is trying to address. The spec is more or less trying to indicate which endpoints are "high traffic".

Reviewer 8: > Servers can also cache this reasonably well on a per-room basis, and the response size is relatively fixed, making it easy to constrain resources on.

This isn't really true since the response heavily depends on user permissions to see rooms.

Reviewer 6: > It's probably fair to consider this endpoint not rate limited for now

@turt2live: to be explicit: this means that you are happy for this to be left as-is?

Reviewer 9: The opposite: the conversation on this thread is leading me to believe that rate limiting should be required given there's no obvious reason for it to be exempt.

Reviewer 6: ok. I agree with that logic :)

#### Request format

```text

> GET /_matrix/client/r0/rooms/{roomID}/hierarchy

Reviewer 9: We should probably call this a `space_hierarchy` or similar to avoid confusion with room upgrade "hierarchies".

Reviewer 9: applies to the federation endpoint too

Reviewer 8: I'd really like to avoid further bike-shedding on the URL.

The only reference in the spec I see to hierarchies is at the [description of room versions](https://matrix.org/docs/spec/#id5):

> There is no implicit ordering or hierarchy to room versions, and their principles are immutable once placed in the specification. Although there is a recommended set of versions, some rooms may benefit from features introduced by other versions. Rooms move between different versions by "upgrading" to the desired version. Due to versions not being ordered or hierarchical, this means a room can "upgrade" from version 2 to version 1, if it is so desired.

I don't really see how this could be confused with this endpoint.

Reviewer 9: It's less so the actual spec definition of room versions and more so how the community and ecosystem views room versions. No matter how hard we try to communicate how room versions work, including documenting all their nuances, people still see them as rolling improvements with an "old room" and a "new room", creating a hierarchy. 

I'm very interested in not bikeshedding this to death, but I do think that the word "space" needs to be in the endpoint. This doesn't require re-implementing the MSC or changing how clients use the unstable prefix, so should be safe to change.

Reviewer 8: I would be interesting in hearing the opinions of others, there was already a bikeshed about this (https://github.com/matrix-org/matrix-doc/pull/2946#discussion_r678442008 -- I don't know if that will load well because of GitHub).

> it kinda feels a bit silly that we are asking for `/spaces` when, for all we know, none of the rooms in the result will actually be spaces. But that's bikeshedding and I'm not sure I have any better ideas.

`/children` and `/hierarchy` were the other suggestions we came up with and settled on `/hierarchy` due it not returning strictly children.

Reviewer 9: The rationale for the endpoint naming doesn't really read as something that was fully considered to me. I'm still not happy with the naming here, but it's also not a blocking concern.

```

Query Parameters:

* **`suggested_only`**: Optional. If `true`, return only child events and rooms
  where the `m.space.child` event has `suggested: true`.  Must be a  boolean,
  defaults to `false`.

  This applies transitively, i.e. if a `suggested_only` is `true` and a space is
  not suggested then it should not be searched for children. The inverse is also
  true, if a space is suggested, but a child of that space is not then the child
  should not be included.
* **`limit`**: Optional: a client-defined limit to the maximum

>   number of rooms to return per page. Must be a non-negative integer.

Reviewer 9: Nit: 
```suggestion
  number of rooms to return per page. Must an integer greater than zero.
```

Reviewer 6: it can be zero though?

Reviewer 8: Including a limit of 0 seems to be not useful (it would be an empty response). The code allows this right now though, but we shouldn't.

  Server implementations should impose a maximum value to avoid resource
  exhaustion.
* **`max_depth`**: Optional: The maximum depth in the tree (from the root room)
  to return. The deepest depth returned will not include children events. Defaults

>   to no-limit. Must be a non-negative integer.

Reviewer 9: Nit:
```suggestion
  to no-limit. Must be an integer greater than zero.
```

Reviewer 11: I'd say a value of 0 would make sense for the depth, it would just return the top level room, wouldn't it?

  Server implementations may wish to impose a maximum value to avoid resource
  exhaustion.
* **`from`**: Optional. Pagination token given to retrieve the next set of rooms.

  Note that if a pagination token is provided, then the parameters given for
  `suggested_only` and `max_depth` must be the same.

#### Response Format

* **`rooms`**: `[object]` For each room/space, starting with the root room, a
  summary of that room. The fields are the same as those returned by
  `/publicRooms` (see
  [spec](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-client-r0-publicrooms)),
  with the addition of:
  * **`room_type`**: the value of the `m.type` field from the room's
    `m.room.create` event, if any.
  * **`children_state`**: The `m.space.child` events of the room. For each event,
    only the following fields are included<sup id="a1">[1](#f1)</sup>:

>     `type`, `state_key`, `content`, `room_id`, `sender`,  with the addition of:

Reviewer 9: can we put the words "stripped state" somewhere in here to pull in all the behaviour of [MSC3173](https://github.com/matrix-org/matrix-doc/pull/3173) ?

Reviewer 8: Yeah, it's in the footnote but I'm not really sure why that's a footnote. 🤷 Probably left over from rewritings!

>     * **`origin_server_ts`**: This is required for sorting of rooms as specified

Reviewer 9: origin timestamp of what?

Reviewer 9: this is also not in the example below

Reviewer 8: I think this is fairly clear:

> For each event, only the following fields are included [...], with the addition of: `origin_server_ts`

Can you provide more info on how you would like this to be clarified?

(I've fixed the example already.)

Reviewer 9: The indentation tripped me up on this :( 

I'm happy to blame Markdown on this though.

behavior of other room endpoints (e.g.
[`/_matrix/client/r0/rooms/{roomID}/aliases`](https://matrix.org/docs/spec/client_server/latest#get-matrix-client-r0-rooms-roomid-aliases))
to not divulge that a room exists which the user doesn't have permission to view.

An HTTP response with a status code of 400 and an error code of `M_INVALID_PARAM`
should be returned if the `from` token provided is unknown to the server or if
the `suggested_only` or `max_depth` parameters are modified during pagination.

#### Server behaviour

The server should generate the response as discussed above, by doing a depth-first
search (starting at the "root" room) for any `m.space.child` events. Any
`m.space.child` with an invalid `via` are discarded (invalid is defined as in
[MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772): missing, not an
array or an empty array).

In the case of the homeserver not having access to the state of a room, the
server-server API (see below) can be used to query for this information over
federation from one of the servers provided in the `via` key of the
`m.space.child` event. It is recommended to cache the federation response for a

> period of time and to prefer local data over data returned over federation.

Reviewer 8: If you have Space A with Room B in it. HS1 is in Room B, but not Space A. When HS1 queries HS2 for Space A over federation the contents of Room B will be returned. Those should be discarded and calculated directly by HS1.

> period of time and to prefer local data over data returned over federation.

Reviewer 12: Aha, that makes sense.  Can we make that more explicit?  Maybe we can change the wording to

```suggestion
period of time.  The federation results may contain information on a room
that the requesting server is already participating in; the requesting server
should use its local data for such rooms rather than the data returned over
federation.
```

(which is significantly more wordy, but I'm not sure how to make it less wordy)

When the current response page is full, the current state should be persisted
and a pagination token should be generated (if there is more data to return).
If the client does not request the next page after a short period of time the

> persisted data may be discarded to limit resource usage.

Reviewer 8: That's fair, I don't know if we should be requiring that as part of the MSC though, it seems a little deep in the weeds of how to implement it. This was mostly included to have some logic on protecting the server from resource exhaustion.

> persisted data may be discarded to limit resource usage.

Reviewer 12: How about just something like "To prevent resource exhaustion, the server may expire persisted data that it deems to be stale."

> persisted data may be discarded to limit resource usage.

Reviewer 8: I'll add something like that, thank you! :)

> The persisted state will includes:

Reviewer 6: ```suggestion
The persisted state will include:
```

easily known the room summary might have changed).

Since the server-server API does not know the requesting user, the response should
divulge information based on if any member of the requesting server could join
the room. The requesting server is trusted to properly filter this information
using the `world_readable`, `join_rules`, and `allowed_room_ids` fields from the
response.

If the target server is not a member of some children rooms (so would have to send
another request over federation to inspect them), no attempt is made to recurse
into them. They are simply omitted from the `children` key of the response.
(Although they will still appear in the `children_state`key of the `room`.)

Similarly, if a server-set limit on the size of the response is reached, additional
rooms are not added to the response and can be queried individually.

#### Request format

```text
GET /_matrix/federation/v1/hierarchy/{roomID}
```

Query Parameters:

* **`suggested_only`**: The same as the Client-Server API.

#### Response format

The response format is similar to the Client-Server API:

* **`room`**: `[object]` The summary of the requested room, as given in the

>   Client-Server API response.

Reviewer 11: I think this should also include the `allowed_room_ids`, since you could have a restricted space, that needs room membership to join it (and at least one of those rooms exists on the requesting server).

Reviewer 8: The initial request does not ever get sent over federation. I'm not sure it actually says this anywhere above though. I need to double check.

Reviewer 11: I would really like to be able to do the first request over federation for https://github.com/matrix-org/matrix-doc/pull/3266. And I think this could be useful for this MSC alone as well.

Reviewer 8: This is mentioned above: https://github.com/matrix-org/matrix-doc/blob/kegan/spaces-summary/proposals/2946-spaces-summary.md#errors

> An HTTP response with a status code of 403 and an error code of `M_FORBIDDEN` should be returned if the user doesn't have permission to view/peek the root room.

Reviewer 11: Let's say we have space A. User x and y are on server S. The space A is on server U. You need to be member of room B to join the space A. x is member of room B, y isn't. If we now request a summary of the space A over federation, we won't get the room ids required to join A. So if y requested it, we can't know, if we should return the response to them or not.

Am I missing something?

    `m.room.create` event, if any.
  * **`allowed_room_ids`**: A list of room IDs which give access to this room per
    [MSC3083](https://github.com/matrix-org/matrix-doc/pull/3083).<sup id="a2">[2](#f2)</sup>
* **`inaccessible_children`**: Optional `[string]`. A list of room IDs which are
  children of the requested room, but are inaccessible to the requesting server.
  The requesting server should not attempt to request information about them

>   from other servers.

Reviewer 9: for clarity: is this a SHOULD or a should? Difference being whether or not it ends up in the spec verbatim as a recommendation rather than a requirement: it feels important to let servers verify this bit of information with another server to ensure the responding server isn't misleading the requesting server.

Reviewer 8: I think "SHOULD" -- it is a waste of resources to attempt to further request those room IDs from other servers.

> it feels important to let servers verify this bit of information with another server to ensure the responding server isn't misleading the requesting server.

This could be said about pretty much everything from this federation API. The information isn't really verifiable from the requesting server.

Reviewer 9: It's true from a federation perspective, but there's always avenues a server can take to verify the data it receives. We haven't yet made a suggestion in the spec for a server to *not* try other servers, though we also haven't put in a recommendation that they *should* do so. 

It's a bit of a philosophical argument about whether or not it's fair for the spec to say a server shouldn't try another server or not. For Matrix, we've typically leant more on the side of letting servers make bad decisions with respect to resource consumption but provide a way for them to improve (such as shortcuts in state res) - for this case, other servers might still have the information but it's highly unlikely, thus the requesting server would be wasting it's time. 

Rewording the sentence puts the onus on the requesting server to decide if it wants to waste resources or not, but still maintains the recommendation to not bother:

```suggestion
  Other servers are unlikely to have information about them as well, thus the
  requesting server can consider the rooms inaccessible from everywhere.
```

  This is used to differentiate between rooms which the requesting server does
  not have access to from those that the target server cannot include in the
  response (which will simply be missing in the response).

#### Example request:

```jsonc
GET /_matrix/federation/v1/hierarchy/{roomID}?
    suggested_only=true
```

#### Errors:

> An HTTP response with a status code of 404 is returned if the target server is

Reviewer 9: ```suggestion
An `M_NOT_FOUND` error with a status code of 404 is returned if the target server is
```

> of the characters in `order` values. Rooms with no `order` come last, in
> ascending numeric order of the `origin_server_ts` of their `m.room.create`
> events, or ascending lexicographic order of their `room_id`s in case of equal
> `origin_server_ts`. `order`s which are not strings, or do not consist solely
> of ascii characters in the range `\x20` (space) to `\x7F` (~), or consist of
> more than 50 characters, are forbidden and the field should be ignored if
> received.

Unfortunately there are situations when a homeserver comes across a reference to
a child room that is unknown to it and must decide the ordering. Without being
able to see the `m.room.create` event (which it might not have permission to see)
no proper ordering can be given.

Consider the following case of a space with 3 child rooms:

```
         Space A
           |
  +--------+--------+
  |        |        |
Room B   Room C   Room D
```

Space A, Room B, and Room C are on HS1, while Room D is on HS2. HS1 has no users
in Room D (and thus has no state from it). Room B, C, and D do not have an
`order` field set (and default to using the ordering rules above).

> When a user asks HS1 for the space summary with a `max_rooms_per_space` equal to

Reviewer 9: Where does `max_rooms_per_space` come from?

Reviewer 8: This field no longer exists, I've updated this section.

> # MSC2946: Spaces Summary

Reviewer 13: I'm thinking about ways to speed up "space exploration" performance and one way would be to just return the room ids of all rooms and subspaces of a space so the client can lazily fetch them as it needs

That way clients can almost instantly present *something* to the user while waiting for more detailed information to arrive over federation

But maybe clients can already do this by calling other endpoints

Reviewer 13: Thoughts on this? Last time I tested space exploration on Synapse it was very slow and this could improve it.

Reviewer 8: It has been significantly improved recently.

I suspect the overhead of lazily fetching each room would be much slower in the case of a user who is not in many of the rooms in a space.

It might be faster in the case of a user who is in all rooms in a space.

Part of way the API is intensive is the recursive nature of it, I'm unsure if you're suggesting *not* doing that in your comment. If so, that's already possible by giving the `max_depth` parameter.

  to no-limit. Must be a non-negative integer.

  Server implementations may wish to impose a maximum value to avoid resource
  exhaustion.
* **`from`**: Optional. Pagination token given to retrieve the next set of rooms.

  Note that if a pagination token is provided, then the parameters given for
  `suggested_only` and `max_depth` must be the same.

#### Response Format

* **`rooms`**: `[object]` For each room/space, starting with the root room, a
  summary of that room. The fields are the same as those returned by
  `/publicRooms` (see
  [spec](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-client-r0-publicrooms)),
  with the addition of:
  * **`room_type`**: the value of the `m.type` field from the room's
    `m.room.create` event, if any.
  * **`children_state`**: The `m.space.child` events of the room. For each event,
    only the following fields are included<sup id="a1">[1](#f1)</sup>:
    `type`, `state_key`, `content`, `room_id`, `sender`,  with the addition of:
    * **`origin_server_ts`**: This is required for sorting of rooms as specified
      below.
* **`next_batch`**: Optional `string`. The token to supply in the `from` param

>   of the next `/spaces` request in order to request more rooms. If this is absent,

Reviewer: ```suggestion
  of the next `/hierarchy` request in order to request more rooms. If this is absent,
```

  Server implementations should impose a maximum value to avoid resource
  exhaustion.
* **`max_depth`**: Optional: The maximum depth in the tree (from the root room)
  to return. The deepest depth returned will not include children events. Defaults
  to no-limit. Must be a non-negative integer.

  Server implementations may wish to impose a maximum value to avoid resource
  exhaustion.
* **`from`**: Optional. Pagination token given to retrieve the next set of rooms.

  Note that if a pagination token is provided, then the parameters given for
  `suggested_only` and `max_depth` must be the same.

#### Response Format

* **`rooms`**: `[object]` For each room/space, starting with the root room, a
  summary of that room. The fields are the same as those returned by
  `/publicRooms` (see
  [spec](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-client-r0-publicrooms)),
  with the addition of:
  * **`room_type`**: the value of the `m.type` field from the room's
    `m.room.create` event, if any.
  * **`children_state`**: The `m.space.child` events of the room. For each event,
    only the following fields are included<sup id="a1">[1](#f1)</sup>:

>     `type`, `state_key`, `content`, `room_id`, `sender`,  with the addition of:

Reviewer 8: I wonder if `room_id` is actually useful here after nesting `children_state` under each entry in `rooms`? It should just match the parent's `room_id` field.

Reviewer 8: I've removed this as it is redundant.

    * **`origin_server_ts`**: This is required for sorting of rooms as specified

>       below.

Reviewer 6: Unless I'm misreading the code, the current implementation in Synapse, doesn't appear to follow this:  https://github.com/matrix-org/synapse/blob/v1.41.0/synapse/handlers/room_summary.py#L648-L652. See also https://github.com/matrix-org/matrix-doc/issues/877.

Reviewer 8: Good shout, we're not doing this properly. I've put up a fix for Synapse: matrix-org/synapse#10760.

I'm not sure if this invalidates your comment on #877 or not.

Reviewer 6: I think it does. THanks!
